classes/factor.php

```diff
@@ -0,0 +1,132 @@
+ * @license     http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class factor extends object_factor_base {
+
+    /**
+     * Exemption Factor implementation.
+     *
+     * @param stdClass $user the user to check against.
+     * @return array
+     */
+    public function get_all_user_factors($user) {
+        return $this->get_singleton_user_factor($user);
+    }
+
+    /**
+     * Exemption Factor implementation.
+     *
+     * {@inheritDoc}
+     */
+    public function has_input() {
+        return false;
+    }
+
+    /**
+     * Exemption Factor implementation.
+     *
+     * {@inheritDoc}
+     */
+    public function get_state() {
+        global $DB, $USER;
+
+        // As long as the user has an exemption record that hasnt expired yet, they pass.
+        $select = 'userid = ? AND expiry > ?';
+        $exempt = $DB->record_exists_select('factor_exemption', $select, [$USER->id, time()]);
+
+        return $exempt ? \tool_mfa\plugininfo\factor::STATE_PASS : \tool_mfa\plugininfo\factor::STATE_NEUTRAL;
+    }
+
+
+    /**
+     * Exemption factor implementation.
+     *
+     * @param \stdClass $user
+     */
+    public function possible_states($user) {
+        // Exemption can only be neutral or pass.
+        return [
+            \tool_mfa\plugininfo\factor::STATE_PASS,
+            \tool_mfa\plugininfo\factor::STATE_NEUTRAL,
+        ];
+    }
+
+    /**
+     * Add an exemption for a user.
+     *
+     * @param \stdClass $user
+     */
+    public static function add_exemption(\stdClass $user) {
+        global $DB;
+
+        // We don't need to handle logic for dealing with anything here except inserts.
+        // Duplicate records do not cause any issues.
+        $duration = get_config('factor_exemption', 'duration');
+        $record = [
+            'userid' => $user->id,
+            'expiry' => time() + (int) $duration,
+            'timecreated' => time()
+        ];
+        $DB->insert_record('factor_exemption', $record);
+
+        $event = \factor_exemption\event\exemption_added::exemption_added_event($user);
+        $event->trigger();
+    }
+
+    /**
+     * Extend an exemption for the user.
+     * This should be used instead of creating a new extension to avoid cluttering the table with duplicate records.
+     *
+     * @param int $eid the exemption record id.
+     */
+    public static function extend_exemption(int $eid) {
+        global $DB;
+
+        $duration = get_config('factor_exemption', 'duration');
+        $DB->set_field('factor_exemption', 'expiry', time() + (int) $duration, ['id' => $eid]);
+
+        $event = \factor_exemption\event\exemption_extended::exemption_extended_event($eid);
+        $event->trigger();
+
+    }
+
+    /**
+     * Delete an exemption for the user.
+     * This doesn't do a real delete,  just sets the expiry in the past.
+     *
+     * @param int $eid the exemption record id.
+     */
+    public static function delete_exemption(int $eid) {
```

Comment:
Since this doesn't actually delete them, might be worth renaming it to something like `expire_exemption` to be clearer